writer.go

```diff
@@ -0,0 +1,174 @@
+package sequencefile
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/rand"
+	"encoding/binary"
+	"io"
+
+	"github.com/golang/snappy"
+)
+
+type Writer struct {
+	Header        Header
+	writer        io.Writer
+	sinceLastSync int
+	compressor    compressor
+}
+
+func NewWriter(w io.Writer) *Writer {
+	header := Header{
+		Version:        6,
+		KeyClassName:   "org.apache.hadoop.io.BytesWritable",
+		ValueClassName: "org.apache.hadoop.io.BytesWritable",
+		Compression:    NoCompression,
+	}
+	return &Writer{writer: w, Header: header, sinceLastSync: 0}
+}
+
+func (w *Writer) Write(bytes []byte) (int, error) {
+	return w.writer.Write(bytes)
+}
+
+func (w *Writer) writeSyncMarker() (int, error) {
+	if w.Header.SyncMarker == nil {
+		syncMarkerBytes := make([]byte, SyncSize)
+		_, err := rand.Read(syncMarkerBytes)
+		if err != nil {
+			return 0, err
+		}
+		w.Header.SyncMarker = syncMarkerBytes
+	}
+
+	w.sinceLastSync = 0
+	return w.writer.Write(w.Header.SyncMarker)
+}
+
+func (w *Writer) sync() (int, error) {
+	if w.sinceLastSync > 1024*8 {
+		totalwritten := 0
+		written, err := w.writer.Write([]byte{0xff, 0xff, 0xff, 0xff})
+		totalwritten += written
+		if err != nil {
+			return totalwritten, err
+		}
+		written, err = w.writeSyncMarker()
+		totalwritten += written
+		return totalwritten, err
+	}
+	return 0, nil
+}
+
+func (w *Writer) Append(key []byte, value []byte) (int, error) {
+	// TODO: if we haven't written the header yet, should we error, or just silently write the header?
+	var written int
+	var err error
+	totalwritten := 0
+
+	written, err = w.sync()
+	totalwritten += written
+	if err != nil {
+		return totalwritten, err
+	}
+
+	keylength := len(key)
+	keylengthbytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(keylengthbytes, uint32(keylength))
+
+	if w.Header.Compression == RecordCompression {
+		value, err = w.compress(value)
+		if err != nil {
+			return totalwritten, err
+		}
+	}
+
+	recordlength := keylength + len(value)
+	recordlengthbytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(recordlengthbytes, uint32(recordlength))
+
+	written, err = w.writer.Write(recordlengthbytes)
+	totalwritten += written
+	if err != nil {
+		return totalwritten, err
+	}
+
+	written, err = w.writer.Write(keylengthbytes)
+	totalwritten += written
+	if err != nil {
+		return totalwritten, err
+	}
+
+	written, err = w.writer.Write(key)
+	totalwritten += written
+	if err != nil {
+		return totalwritten, err
+	}
+
+	written, err = w.writer.Write(value)
+	totalwritten += written
+	if err != nil {
+		return totalwritten, err
+	}
+
+	w.sinceLastSync += totalwritten
+	return totalwritten, nil
+}
+
+func (w *Writer) compress(raw []byte) ([]byte, error) {
+	switch w.Header.CompressionCodec {
+	case GzipCompression:
+		return w.compressGzip(raw)
+	case SnappyCompression:
+		return w.compressSnappy(raw)
+	default:
+		panic("compression set without codec")
+	}
+}
+
+func (w *Writer) compressSnappy(raw []byte) ([]byte, error) {
```

Comment:
yea, I was trying to. I think it's harder because of the framing format. and because blockcompression. It's been a while since I thought about this though so I'll look it over again and think about it and see :)